mate3/api.py

```diff
@@ -164,54 +303,56 @@ def set_value(self, field: Field, value: int, port: int = None):
-        found_devices = []
-        for device, start_register in self._block_information():
-            if field.device != device:
-                continue
-
-            parser = get_parser(device)
-            if hasattr(parser, 'port_number'):
-                structure = parse(device, self.client, start_register, only_fields=[parser.port_number])
-                if port is not None and structure.port_number != port:
-                    continue
-
-                found_devices.append(start_register)
-            else:
-                found_devices.append(start_register)
-
-        # Throw some sensible errors
-        if not found_devices:
-            raise Exception(f"No device of type {field.device.name} found on hub port {port}")
-        if len(found_devices) > 1:
-            if port:
-                raise Exception(
-                    f"Multiple devices of type {field.device.name} found on hub port {port}. "
-                    f"This is very unexpected, please raise an issue in GitHub if you see this, "
-                    f"including details of the device you are trying to set values on."
-                )
-            else:
-                raise Exception(
-                    f"Multiple devices of type {field.device.name} found. Please specify a hub port number "
-                    f"using the `port` parameter."
-                )
-
-        # What register do we need to write to?
-        field_register = found_devices[0] + field.start - 1
-
-        # Split the integer into multiple values according to the field's type.
-        # This will also raise an error if the provided value is to big/small.
-        prepared_values = split_int(value, field.type)
-
-        logger.debug(
-            f"Setting register {field_register} to value {value} "
-            f"(which is a {field.type.__name__} represented as {len(prepared_values)} values: {prepared_values})"
-        )
 
-        # Do the write
-        self.client.write_registers(field_register, prepared_values)
+        # TODO: write in ranges?
+        for device in self._devices.connected_devices:
+            for field_value in device.fields():
+                if field_value.dirty:
+                    if field_value.field.mode not in (Mode.RW, Mode.W):
+                        raise RuntimeError(f"Can't write to field {field_value.name}")
+                    logger.info(f"writing {field_value._value_to_write} to {field_value.name}")
+                    address = device.address + field_value.field.start - 1  # -1 since start is 1-indexed
+                    # TODO: should we check that the device at the given address still has the right port?
+                    registers = field_value.field.to_registers(field_value._value_to_write)
+                    logger.debug(f"Setting register {address} to value {registers}")
+
+                    # # Do the write
+                    self.client.write_registers(address, registers)
+
+
+if __name__ == "__main__":
```

Comment:
This looks like debug code, perhaps this needs to be removed or placed elsewhere such as a script entrypoint?

Reply:
Good point - I actually moved it out to `examples/getting_started.md` so I'll delete it from `api.py`.